src/core_ocean/shared/mpas_ocn_tracer_sediment_mod.F

```diff
@@ -0,0 +1,235 @@
+! !INTERFACE:
+
+ subroutine SEDIMENT_SourceSink(SEDIMENT_indices, SEDIMENT_input, SEDIMENT_forcing, &
+                                SEDIMENT_output, SEDIMENT_diagnostic_fields, &
+                                numLevelsMax, numColumnsMax, numColumns)
+
+! !DESCRIPTION:
+!  Compute time derivatives for tracegas state variables
+!
+! !REVISION HISTORY:
+!  same as module
+
+! !INPUT PARAMETERS:
+
+  type(SEDIMENT_indices_type),     intent(in ) :: SEDIMENT_indices
+  type(SEDIMENT_input_type),       intent(in ) :: SEDIMENT_input
+  type(SEDIMENT_forcing_type),     intent(in ) :: SEDIMENT_forcing
+
+  integer (SEDIMENT_i4) :: numLevelsMax, numColumnsMax, numColumns
+
+! !OUTPUT PARAMETERS:
+
+  type(SEDIMENT_output_type),      intent(inout) :: SEDIMENT_output
+  type(SEDIMENT_diagnostics_type), intent(inout) :: SEDIMENT_diagnostic_fields
+
+!-----------------------------------------------------------------------
+!  local variables
+!-----------------------------------------------------------------------
+
+  integer(SEDIMENT_i4) ::  &
+      column, kmax, k
+
+  integer(SEDIMENT_i4) ::  &
+      sedConcen_ind        ! suspended sediment cocentration index
+
+  real(SEDIMENT_r8), allocatable, dimension(:,:) :: & !kg/m^3
+         sedConcen_loc     ! local copy of model sedConcen
+!---------------------------------------------------------------------------
+!---------------------------------------------------------------------------
+
+!-----------------------------------------------------------------------
+!  initialize  all tendencies to zero
+!-----------------------------------------------------------------------
+
+   SEDIMENT_output%SEDIMENT_tendencies = 0.0_SEDIMENT_r8
+
+!-----------------------------------------------------------------------
+!  allocate local copies of tracers
+!-----------------------------------------------------------------------
+
+   allocate(sedConcen_loc(numLevelsMax,numColumns))
+!-----------------------------------------------------------------------
+!  assign indices.  this is not necessary but results in fewer
+!    differences between original and new code.
+!-----------------------------------------------------------------------
+
+   sedConcen_ind    = SEDIMENT_indices%sedConcen_ind
+
+!-----------------------------------------------------------------------
+!  loop over columns
+!-----------------------------------------------------------------------
+
+   setup_loop: do column = 1, numColumns
+
+   kmax = SEDIMENT_input%number_of_active_levels(column)
+   if (kmax < 1) cycle setup_loop
+
+   do k = 1, kmax
+
+    !---------------------------------------------------------------------------
+    !   create local copies of requisite ecotracers
+    !   treat negative values as zero and apply mask to locals
+    !---------------------------------------------------------------------------
+
+   sedConcen_loc(k,column)      = max(0.0_SEDIMENT_r8, SEDIMENT_input%SEDIMENT_tracers(k,column,sedConcen_ind))
```

Comment:
This is a naive question but can you explain why getting rid of negative values happens here as opposed to elsewhere?

Reply:
I just emulated the codes of DMS and MACROS tracers.

Reply:
we really don't want to act on any sed tracers with negative concentration, but for conservation we don't want to clip the value of the actual prognostic field, so it's done using a temporary.